### PR TITLE
Allow implicit cast between (GenICam) int, float, and enum

### LINKED ISSUE #98: Weaken restriction on `pValue` handling
Currently, we expect each `pValue` element has the same interface as the parent node. e.g. `pNode` of `FloatNode` should have `IFloat` interface. But in some cameras, some `pNode` may have a different interface as a parent node.

So we decided to weaken the restriction so that `pNode` is allowed to have a different interface if there is a int-float conversion between two interface. e.g. `pNode` of `FloatNode` can have `IInteger`/`IFloat`/`IEnumeration`, while can NOT have `IString`/`ICommand`/`IBoolean`/`IPort`/`IRegister`/`ICategory`/`IPort`/`ISelector`.

### Implementation
Rewrite usage of `expect_kind` [here](https://github.com/cameleon-rs/cameleon/blob/cfe66d17c65fc7e345e791840188d4c38189b4ef/genapi/src/ivalue.rs#L138-L183).

### LINKED ISSUE #168: Trouble with "AcquisitionStart" command
- **What did you expect to happen? What did actually happen?**
The stream.rs example is failing to stream from my usb3vision camera. It results in an error propagated up out of the library:
```
2023-08-15T03:43:26.701063Z ERROR start_streaming{cap=3 camera=CameraInfo { vendor_name: "Toshiba-Teli", model_name: "BU505MG", serial_number: "xxxxxxx" }}:execute{node="AcquisitionStart"}: invalid node: the node doesn't implement `IInteger`
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: GenApiError(InvalidNode("the node doesn't implement `IInteger`"))', cameleon/examples/stream.rs:32:48
```

- **How to reproduce the issue?**
Plug in the camera, run the example.

### Meta
- Which manufacturer's camera did you use?
Toshiba Teli
- What is the model name of the camera?
BU505MG
- Which `Cameleon` version are you using?
I'm using current head of main (513cb2534c3aa0bff45f509b50e62e64fbe2ca56, version 0.1.11)

### Investigation
TL;DR: With a tiny hack in Cameleon, I can get it streaming. "AcquisitionStart" command is expecting an enum argument, not an int.

I'm not sure if either my camera is non-conformant, or if there is a missing detail in the way commands are executed in Cameleon. The error originates from the start_streaming method when attempting to execute the AcquisitionStart command. If I break it into two lines, it dies on the execute line:
```
        let acq_start_node = &expect_node!(&ctxt, "AcquisitionStart", as_command);
        acq_start_node.execute(&mut ctxt)?; // <----- dies here
```

Since Cameleon is trying to send an int, let's take a look at my cameras XML. The relevant snippet:

```
...
        <Command Name="AcquisitionStart">
                <ToolTip>Starts the Acquisition of the device.</ToolTip>
                <Description>Starts the Acquisition of the device. The number of frames captured is specified by AcquisitionMode.</Description>
                <DisplayName>Acquisition Start</DisplayName>
                <Visibility>Beginner</Visibility>
                <pIsImplemented>AcquisitionCommandImplemented</pIsImplemented>
                <pIsAvailable>AcquisitionCommandActive</pIsAvailable>
                <!-- <pIsAvailable>AcquisitionStartActive</pIsAvailable> -->
                <pValue>FeatureCategoryBlock3AcquisitionCommandValue</pValue>
                <pCommandValue>AcquisitionMode</pCommandValue>
        </Command>

        <Command Name="AcquisitionStop">
                <ToolTip>Stops the Acquisition of the device at the end of the current Frame.</ToolTip>
                <Description>Stops the Acquisition of the device at the end of the current Frame. It is mainly used when AcquisitionMode is Continuous but can be used in any acquisition mode.</Description>
                <DisplayName>Acquisition Stop</DisplayName>
                <Visibility>Beginner</Visibility>
                <pIsImplemented>AcquisitionCommandImplemented</pIsImplemented>
                <pIsAvailable>AcquisitionCommandActive</pIsAvailable>
                <pValue>FeatureCategoryBlock3AcquisitionCommandValue</pValue>
                <CommandValue>1</CommandValue>
        </Command>
...
```

so whereas the AcquisitionStop value is an integer, the AcquisitionStart command refers to the following enumeration:

```
        <Enumeration Name="AcquisitionMode">
                <ToolTip>Sets the acquisition mode of the device.</ToolTip>
                <Description>Sets the acquisition mode of the device. It defines mainly the number of frames to capture during an acquisition and the way the acquisition stops.</Description>
                <DisplayName>Acquisition Mode</DisplayName>
                <Visibility>Beginner</Visibility>
                <pIsImplemented>AcquisitionCommandImplemented</pIsImplemented>
                <pIsAvailable>AcquisitionCommandActive</pIsAvailable>

        <!--    <pIsAvailable>AcquisitionStartActive</pIsAvailable> -->
                <EnumEntry Name="ImageBufferRead">
                        <pIsImplemented>AcquisitionCommandListOfElement_10</pIsImplemented>
                        <Value>0xA</Value>
                </EnumEntry>
                <EnumEntry Name="MultiFrame">
                        <pIsImplemented>AcquisitionCommandListOfElement_9</pIsImplemented>
                        <Value>0x9</Value>
                </EnumEntry>
                <EnumEntry Name="Continuous">
                        <pIsImplemented>AcquisitionCommandListOfElement_8</pIsImplemented>
                        <Value>0x8</Value>
                </EnumEntry>
                <Value>0x8</Value>
        </Enumeration>
```

Based on this, if I add a hack in cameleon:

```
impl ICommand for CommandNode {
    #[tracing::instrument(skip(self, device, store, cx),
                          fields(node = store.name_by_id(self.node_base().id()).unwrap()))]
    fn execute<T: ValueStore, U: CacheStore>(
        &self,
        device: &mut impl Device,
        store: &impl NodeStore,
        cx: &mut ValueCtxt<T, U>,
    ) -> GenApiResult<()> {
        cx.invalidate_cache_by(self.node_base().id());

>>>>>> if self.attr_base.id.0 == 178 {
>>>>>>   self.value.set_value(0x08, device, store, cx)
>>>>>> }else{
            let value = self.command_value.value(device, store, cx)?;
            self.value.set_value(value, device, store, cx)
>>>>>> }
    }
...
```

where **0x08** is based on the corresponding entry in the XML for "AcquisitionMode" "Continuous", then it can stream **successfully**!

So, I'm not familiar enough with GenICam etc to know what the right solution is here. Is this a bug, or a weird camera? What do you think the correct solution is?

(Thanks for a cool library!)

### PR DESCRIPTION
**Pull request by Y-Nak**
- [x] Check `test_all.sh` is passed.
- [x] Add fix #168, fix #98
- [x] Fill out `## Changelog` section. If the change is for only internal use, please write `None` to the section.

This change allows an implicit casting between int, float, and enum types when the GenApi context requires this kind of casting. 
NOTE: This change doesn't affect the public API.

### CODE REVIEW

**Comment by stevenlovegrove** [comment]
I just tested this and it works great! The streaming example now performs as expected. Thanks so much for fixing this!

**Comment by Y-Nak** [comment]
@stevenlovegrove Thanks for your testing! I changed other parts to avoid similar errors relating to implicit casting. Would you test the code in your environment? Thanks!

**Review by bschwind — APPROVED** [review]
Looks good and the tests pass on the genicam camera I have (3 tests fail, but they were also failing in `main` so it should be a separate issue)

I'm less sure about the code in `ivalue` because it's quite macro heavy, I just left a few small questions overall.

File: genapi/src/elem_type.rs [hunk 1]
```diff
@@ -310,12 +310,10 @@ impl RegPIndex {
         store: &impl NodeStore,
         cx: &mut ValueCtxt<T, U>,
     ) -> GenApiResult<i64> {
-        let base = self
-            .p_index
-            .expect_iinteger_kind(store)?
-            .value(device, store, cx)?;
+        let base = self.p_index.value(device, store, cx)?;
```
[the remark this answers is not in the document]
Comment (Y-Nak) [reply genapi/src/elem_type.rs:313]: I felt it'd be a little dangerous to set an arbitrary value to an enum node. But, since `IValue` is just used internally to cast a type by following the given GenICam XML, I changed it to allow setting an arbitrary value to an enum node.

File: genapi/src/ivalue.rs [hunk 2]
```diff
@@ -132,55 +134,228 @@ macro_rules! impl_ivalue_for_vid {
...
+        value: i64,
+        device: &mut impl Device,
+        store: &impl NodeStore,
+        cx: &mut ValueCtxt<U, S>,
+    ) -> GenApiResult<()> {
+        if let Some(i) = self.as_iinteger_kind(store) {
+            i.set_value(value, device, store, cx)
+        } else if let Some(f) = self.as_ifloat_kind(store) {
+            f.set_value(value as f64, device, store, cx)
+        } else {
```
[the remark this answers is not in the document]
Comment (Y-Nak) [reply genapi/src/ivalue.rs:211]: I felt it'd be a little dangerous to set an arbitrary value to an enum node. But, since `IValue` is just used internally to cast a type by following the given GenICam XML, I changed it to allow setting an arbitrary value to an enum node.

**Review by bschwind — APPROVED** [review]
LGTM!

**Comment by Y-Nak** [comment]
@bschwind Thanks for your review!

**Comment by bschwind** [comment]
Thanks for your work!
